pandas/core/index.py

```diff
@@ -391,6 +395,11 @@ def _shallow_copy(self, values=None, infer=False, **kwargs):
 
         if infer:
             attributes['copy'] = False
+            if self._infer_as_myclass:
```

Comment:
To achieve below result. ``.loc`` against ``DatetimeInde`` internally converts the index to ``Int64Index``. Thus, there should be a logic to revert to ``DatetimeIndex``. 

```
pd.Index([1., 2., 3.])._shallow_copy([1, 2, 3], infer=True)
# Int64Index([1, 2, 3], dtype='int64')

idx = pd.date_range('2011-01-01', freq='D', periods=3)
idx._shallow_copy(idx.asi8, infer=True)
# DatetimeIndex(['2011-01-01', '2011-01-02', '2011-01-03'], dtype='datetime64[ns]', freq='D')
```

In current master, the below case results in:

```
idx._shallow_copy(idx.asi8, infer=True)
Int64Index([1293840000000000000, 1293926400000000000, 1294012800000000000], dtype='int64')
```